backend/select_utils.ml

```diff
@@ -116,8 +106,29 @@ let print_traps ppf traps =
   in
   Format.fprintf ppf "(%a)" print_traps traps
 
-let set_traps nfail traps_ref base_traps exit_traps =
-  let traps = combine_traps base_traps exit_traps in
+let rec set_traps env nfail traps_ref base_traps exit_traps =
```

Comment:
Could we try implementing the "proper" solution as well, as it seems like it might not be too much code, and easier to understand?

Reply:
I wrote this PR before #3768, it might not be necessary now. I'll check and update the PR as needed.

Reply:
It is still necessary, as `Cfg_simplify` (which is supposed to handle the removal of dead exception handlers) runs too late: `Select_utils.Stack_offset_and_exn.update_cfg` fails if it encounters any `Pushtrap` operations for which the handler does not exist, and `Cfg_simplify` only runs after that.

Reply:
- Is this "weird piece of code" replacing what we used to have in `env_enter_trywith`?  
- calling `set_traps` recursively with an empty exit list is confusing here. the recurisve call is guaranteed to skip `List.fold_left` in L120-130. and go directly to update `traps_ref` in L132-142. If this is right, factoring out L132-142 into a little helper function and just calling that would make this code easier to understand and no need for `rec`.

Reply:
I have done this refactoring and updated the comment.

Reply:
This is much clearer, thanks!